// @react-page/core/lib/selector/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var DisplaySelectors = __importStar(require("./display"));
var EditableSelectors = __importStar(require("./editable"));
var editable_1 = require("./editable");
exports.editable = editable_1.editable;
exports.editables = editable_1.editables;
var SettingSelectors = __importStar(require("./setting"));
exports.Selectors = {
    Display: DisplaySelectors,
    Setting: SettingSelectors,
    Editable: EditableSelectors,
};
exports.selectors = function (store) { return ({
    editable: function (id) { return editable_1.editable(store.getState(), { id: id }); },
    editables: function (id) { return editable_1.editables(store.getState()); },
}); };
