// @react-page/core/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_dnd_1 = require("react-dnd");
exports.DragSource = react_dnd_1.DragSource;
exports.DropTarget = react_dnd_1.DropTarget;
var actions_1 = require("./actions");
exports.Actions = actions_1.Actions;
var Editable_1 = __importDefault(require("./components/Editable"));
exports.Editable = Editable_1.default;
var Editor_1 = __importStar(require("./Editor"));
exports.Editor = Editor_1.default;
exports.createEmptyState = Editor_1.createEmptyState;
var lazyLoad_1 = __importDefault(require("./helper/lazyLoad"));
exports.lazyLoad = lazyLoad_1.default;
var sanitizeInitialChildren_1 = __importDefault(require("./helper/sanitizeInitialChildren"));
exports.sanitizeInitialChildren = sanitizeInitialChildren_1.default;
var Provider_1 = __importDefault(require("./Provider"));
exports.Provider = Provider_1.default;
var reducer_1 = require("./reducer");
exports.reducer = reducer_1.reducer;
var editable_1 = require("./reducer/editable");
exports.editableReducer = editable_1.editable;
var reduxConnect_1 = require("./reduxConnect");
exports.connect = reduxConnect_1.connect;
exports.ReduxContext = reduxConnect_1.ReduxContext;
exports.ReduxProvider = reduxConnect_1.ReduxProvider;
var selector_1 = require("./selector");
exports.Selectors = selector_1.Selectors;
var i18n_1 = __importDefault(require("./service/i18n"));
exports.i18n = i18n_1.default;
var plugin_1 = __importDefault(require("./service/plugin"));
exports.PluginService = plugin_1.default;
var classes_1 = require("./service/plugin/classes");
exports.ContentPlugin = classes_1.ContentPlugin;
exports.LayoutPlugin = classes_1.LayoutPlugin;
exports.Migration = classes_1.Migration;
exports.Plugin = classes_1.Plugin;
__export(require("./components/hooks"));
var setAllSizesAndOptimize_1 = require("./reducer/editable/helper/setAllSizesAndOptimize");
exports.setAllSizesAndOptimize = setAllSizesAndOptimize_1.setAllSizesAndOptimize;
exports.default = Editor_1.default;
