// @react-page/core/lib/service/plugin/missing.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var ContentMissingComponent = function (props) { return (React.createElement("div", { style: {
        backgroundColor: 'red',
        padding: '8px',
        border: '1px solid black',
        margin: '2px',
        overflowX: 'scroll',
    } },
    "The requested content plugin could not be found.",
    React.createElement("button", { onClick: props.remove }, "Delete Plugin"),
    React.createElement("pre", null, JSON.stringify(props, null, 2)))); };
exports.contentMissing = function (_a) {
    var name = _a.name, version = _a.version;
    return ({
        Component: ContentMissingComponent,
        name: name,
        version: version,
    });
};
var LayoutMissingComponent = function (_a) {
    var children = _a.children, props = __rest(_a, ["children"]);
    return (React.createElement("div", null,
        React.createElement("div", { style: {
                backgroundColor: 'red',
                padding: '8px',
                border: '1px solid black',
                margin: '2px',
                overflowX: 'scroll',
            } },
            "The requested layout plugin could not be found.",
            React.createElement("button", { onClick: props.remove }, "Delete Plugin"),
            React.createElement("pre", null, JSON.stringify(props, null, 2))),
        children));
};
exports.layoutMissing = function (_a) {
    var name = _a.name, version = _a.version;
    return ({
        Component: LayoutMissingComponent,
        name: name,
        version: version,
    });
};
